Implement ADO client

varsanojidan opened the pull request:
Addresses: https://github.com/sourcegraph/sourcegraph/issues/46263
Implement the client for ADO.

Uses the following ADO API for fetching repos by project/org: https://learn.microsoft.com/en-us/rest/api/azure/devops/git/repositories/list?view=azure-devops-rest-7.0&tabs=HTTP

## Test plan
Unit test

varsanojidan commented:
The API that [lists the Repositories](https://learn.microsoft.com/en-us/rest/api/azure/devops/git/repositories/list?view=azure-devops-rest-7.0&tabs=HTTP) doesn't support pagination for some reason, Im looking to see if there is some alternative I can swap it out with.

sashaostrikov commented on internal/extsvc/azuredevops/client.go:
```diff
@@ -0,0 +1,149 @@
+	"github.com/sourcegraph/sourcegraph/internal/ratelimit"
+)
+
+// Client access a ADO via the REST API.
```
maybe?
```suggestion
// Client used to access ADO code host via the REST API.
```

sashaostrikov commented on internal/extsvc/azuredevops/client.go:
```diff
@@ -0,0 +1,149 @@
+
+// Client access a ADO via the REST API.
+type Client struct {
+	// HTTP Client used to communicate with the API
```
🚨 full stop police 🚨
```suggestion
	// HTTP Client used to communicate with the API.
```

sashaostrikov commented on internal/extsvc/azuredevops/client.go:
```diff
@@ -0,0 +1,149 @@
+	// HTTP Client used to communicate with the API
+	httpClient httpcli.Doer
+
+	// Config is the code host connection config for this client
```
🚨 full stop police, officer Nit Pickley, freeze! 🚨
```suggestion
	// Config is the code host connection config for this client.
```

sashaostrikov commented on internal/extsvc/azuredevops/client.go:
```diff
@@ -0,0 +1,149 @@
+	ProjectOrOrgName string
+}
+
+func (c *Client) ListRepositoriesByProjectOrOrg(ctx context.Context, opts ListRepositoriesByProjectOrOrgArgs) (projects *ListRepositoriesResponse, err error) {
```
doesn't seem that named return parameters are used 🤔
should we remove them?

sashaostrikov commented on internal/extsvc/azuredevops/client.go:
```diff
@@ -0,0 +1,149 @@
+		return nil, err
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
```
HTTP codes <200 are technically not errors, does ADO API use it as erroneous codes?

varsanojidan replied:
Thank you for the review brat.
I think its more so that we aren't expecting error codes under 200, and if we get one something went wrong. What do you think?

sashaostrikov replied:
in this sense yep, I agree. SGTM!

sashaostrikov commented on internal/extsvc/azuredevops/client.go:
```diff
@@ -0,0 +1,149 @@
+	IsDisabled bool   `json:"isDisabled"`
+}
+
+type Pagination struct {
```
seems unused?

sashaostrikov replied:
as in ADO API lol

varsanojidan replied:
Very unused by ADO :^)

sashaostrikov commented on internal/extsvc/gerrit/client_test.go:
```diff
@@ -10,7 +10,6 @@ import (
 	"testing"
 
 	"github.com/dnaeon/go-vcr/cassette"
-	"github.com/inconshreveable/log15"
```
nice!

sashaostrikov reviewed (approved):
LGTM overall!
Left some comments.

sashaostrikov reviewed (approved):
:shipit:     :shipit:  :shipit:  :shipit::shipit::shipit::shipit:  :shipit::shipit::shipit::shipit:
:shipit::shipit:  :shipit:  :shipit:  :shipit:       :shipit:   
:shipit: :shipit: :shipit:  :shipit:  :shipit:       :shipit::shipit::shipit: 
:shipit:  :shipit::shipit:  :shipit:  :shipit:       :shipit:    
:shipit:     :shipit:  :shipit:  :shipit::shipit::shipit::shipit:  :shipit::shipit::shipit::shipit:
                             

mrnugget commented on internal/extsvc/azuredevops/client_test.go:
```diff
@@ -0,0 +1,96 @@
+	testutil.AssertGolden(t, "testdata/golden/ListProjects.json", *update, resp)
+}
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	os.Exit(m.Run())
+}
```
Nitpick: I'd put that in `main_test.go`

varsanojidan replied:
I don't actually think there is a reason for it tbh, I can just remove it.

mrnugget replied:
It's there to parse the `update` flag so you can update the golden test data

varsanojidan replied:
right..... 🤦 . whoops.

mrnugget commented on internal/extsvc/azuredevops/client.go:
```diff
@@ -0,0 +1,145 @@
+	}
+
+	resp, err := c.httpClient.Do(req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	bs, err := io.ReadAll(resp.Body)
+
```
So I checked out the branch locally but my computer ran out of disk space because of all these newlines before `if err ...`. Can you remove them?

(I'm joking, but also not: we don't put newlines before the `if err != nil{` in the codebase)

varsanojidan replied:
I just like my padding :P, yeah i'll remove it.

varsanojidan commented on internal/extsvc/azuredevops/client.go:
```diff
@@ -0,0 +1,145 @@
+// the provided configuration. If a nil httpClient is provided, http.DefaultClient
+// will be used.
+func NewClient(urn string, config *ADOConnection, httpClient httpcli.Doer) (*Client, error) {
+	u, err := url.Parse("https://dev.azure.com")
```
That was a good question, for Azure Services (cloud) this is always the API.

But you can run Azure Server on-prem (which I didn't spot before when I originally implemented this), so for those deployments we would need to be able to set the URL.

I'll add it to the conf.

https://learn.microsoft.com/en-us/azure/devops/integrate/how-to/call-rest-api?view=azure-devops

varsanojidan commented:
@mrnugget since this is already merged im going to push all of the changes addressing PR comments to https://github.com/sourcegraph/sourcegraph/pull/46646
